### DISCUSSION TITLE
Toga/Beeware and Vulkan

### DISCUSSION THREAD

**Discussion by skiwan** [beeware/toga#2168] [Q&A] [answered]
Hey community

I havent worked with Beeware/Toga yet (apart from looking at the Tutorial)

I am wondering if there is a possibility to work with glfw / vulkan contexts in Toga/Beeware.
Im currently trying to build a cross platform app with python, using vulkan as the render engine and I am wondering if Beeware would be the right tool to use here.

From other questions I understand that Toga doesnt support OpenGL and I assume that its using its own window manager. What would have to be done if one would like to use vulkan to create the gui in Toga (or parts of it) and then use beeware?

**Comment by freakboy3742** [comment, marked as answer]
Toga doesn't currently have any OpenGL support. 

I've never used GLFW before; but based on a quick scan of the documentation, it seems at least plausible that you would be able to write a Toga widget that wraps a GLFW context. Toga doesn't do any window management or compositing on its own - it's a wrapper around whatever tools and libraries are provided by native platforms. If you're able to work out how to get a Cocoa/Winforms/GTK compatible "widget" out of GLFW (something you could add to a normal Cocoa NSView/Winforms Form, or GTK layout), wrapping that in a Toga layer should be reasonably straightforward.

**Reply by skiwan** [reply]
Gotcha.

But the Toga mainWindow is a window right? Similar to what GLFW or SDL would provide I assume.
Or does mainWindow somehow work as a wrapper around whatever native window management the platform offers?

**Reply by mhsmith** [reply]
> Or does mainWindow somehow work as a wrapper around whatever native window management the platform offers?

Yes, that's exactly what it is.
